### DISCUSSION TITLE
How about business logic in stored procedures

### DISCUSSION THREAD

**Discussion by exotfboy** [PostgREST/postgrest#2509] [Q&A] [answered]
Hi:

PostgREST is a wonderful tool I've ever seen for deploying database model to client directly.

As the document said, for an application, the straightforward CRUD operation can use restful api, the authentication can be implemented by RLS, the complex operation can be down by adding function in db and use the rpc api then.

However in this case the database will be the first class citizen in application, which means some business logic is implemented in the store procedure which seems like [anti-pattern](https://stackoverflow.com/questions/484516/arguments-for-against-business-logic-in-stored-procedures).

How to avoid this?

**Comment by wolfgangwalther** [comment, marked as answer]
> How to avoid this?

Don't.

I put **all** my business logic in the database. I deploy PostgreSQL, PostgREST, nginx as a reverse proxy for some details - and a dumb client. No other server components needed.

Don't fall for using RPCs too often. Many things are better done in constraints, triggers, views, policies, computed columns, procedural languages other than plpgsql like plpython, plv8, ...

Just make sure your tooling is up to standard: Unit tests, local, temporary development databases, all code in git, etc...

Embrace the power of PostgreSQL!

**Reply by exotfboy** [reply]
Looks nice, however there always some logic may not be implemented in db, in this case we have to do the programming, then different application components will live in different place(db/git) which means hard to maintain or version controlling.

**Reply by wolfgangwalther** [reply]
> then different application components will live in different place(db/git)

No. You need to manage all your SQL code in git. Git is the only source of truth - not the database.

If you do it like this, you will never have this problem and maintenance will be fairly simple.

**Reply by exotfboy** [reply]
Seems like I got that, all the procedures/triggers/policies should be maintained in git together with the required codes and deployed automatically when the app start without any manual operation.
